### PR TITLE
[DeepSpeed in notebooks] Jupyter + Colab

### PR DESCRIPTION
**Pull request by stas00**
This PR addresses issues raised in https://github.com/huggingface/transformers/issues/10011 when the user tried to use DeepSpeed in a notebook (most likely colab).

This PR:

* forces device and distributed setup init from TrainingArguments explicitly at the beginning of Trainer's `__init__`. This is needed since until now the init was happening as a side effect of someone calling `device` or `n_gpus`, which doesn't happen if someone runs their own version of `Trainer` w/ deepspeed - which is the case with notebooks - so we are missing out on DeepSpeed init and things weren't working. Let's do it explicitly, and not as a side-effect, so everything is loud and clear.
* sets up `self.local_rank` based on LOCAL_RANK env var under deepspeed to save users a hassle - deepspeed must have `local_rank>-1`. I guess the fake launcher env setup could be folded into the init as well, but then the user loses control over the port number and they may need to edit it, so for now leaving it outside - but will ask deepspeed to provide a wrapper function to make it easy for the user. perhaps once the wrapper is available it could be automated completely. Alternatively, if they make `mpi4py` a dependency then fake launcher env setup won't be needed at all.
* documents how to run DeepSpeed in the notebook env
* adds a test that mocks a notebook environment and runs deepspeed w/o a launcher

I may wait to hear about the follow up to https://github.com/microsoft/DeepSpeed/issues/748 to merge this, if it looks quick, but otherwise I will revise the setup doc in a future PR. The main changes of this PR besides the doc are required anyway.

@sgugger

### CODE REVIEW

File: src/transformers/training_args.py [hunk 1]
```diff
@@ -556,6 +556,11 @@ def _setup_devices(self) -> "torch.device":
             # python -m torch.distributed.launch --nproc_per_node=2 ./program.py
             from .integrations import is_deepspeed_available
 
+            # workaround for setups like notebooks where the launcher can't be used,
+            # but deepspeed requires a dist env.
+            if self.local_rank == -1:
+                self.local_rank = 0
```
Comment (sgugger) [inline src/transformers/training_args.py:562]: Could we at least check the env variable `LOCAL_RANK` exists and is 0 before doing this (which will be the case if the user follow the instructions)? Always changing this makes me a bit nervous.

**Review by sgugger — APPROVED** [review]
Nice new doc! Thanks for adding this.

File: src/transformers/training_args.py (hunk 1 above)
Comment (stas00) [reply src/transformers/training_args.py]: Yes, of course! Done.

**Review by ngoquanghuy99 — COMMENTED** [review]
Thank you so much for creating this!

**Pull request merged**